pkg/cloud/services/compute/instance.go

```diff
@@ -596,61 +530,148 @@ func (s *Service) DeleteInstance(object runtime.Object, instanceName string) err
 				return err
 			}
-			if len(trunkInfo) == 1 {
-				err = util.PollImmediate(RetryIntervalTrunkDelete, TimeoutTrunkDelete, func() (bool, error) {
-					if err := trunks.Delete(s.networkClient, trunkInfo[0].ID).ExtractErr(); err != nil {
-						if capoerrors.IsRetryable(err) {
-							return false, nil
-						}
-						return false, err
-					}
-					return true, nil
-				})
-				if err != nil {
-					return fmt.Errorf("error deleting the trunk %v", trunkInfo[0].ID)
-				}
+		}
+
+		if err = deletePort(s, object, port.PortID); err != nil {
+			return err
+		}
+	}
+
+	return deleteInstance(s, object, instance.ID)
+}
+
+func deletePort(is *Service, object runtime.Object, portID string) error {
+	port, err := ports.Get(is.networkClient, portID).Extract()
+	if err != nil {
+		if capoerrors.IsNotFound(err) {
+			return nil
+		}
+	}
+
+	err = util.PollImmediate(RetryIntervalPortDelete, TimeoutPortDelete, func() (bool, error) {
+		err := ports.Delete(is.networkClient, port.ID).ExtractErr()
+		if err != nil {
+			if capoerrors.IsRetryable(err) {
+				return false, nil
 			}
+			return false, err
 		}
+		return true, nil
+	})
+	if err != nil {
+		record.Warnf(object, "FailedDeletePort", "Failed to delete port %s with id %s: %v", port.Name, port.ID, err)
+		return err
+	}
 
-		// delete port
-		err = util.PollImmediate(RetryIntervalPortDelete, TimeoutPortDelete, func() (bool, error) {
-			err := ports.Delete(s.networkClient, port.PortID).ExtractErr()
-			if err != nil {
-				if capoerrors.IsRetryable(err) {
-					return false, nil
-				}
-				return false, err
+	record.Eventf(object, "SuccessfulDeletePort", "Deleted port %s with id %s", port.Name, port.ID)
+	return nil
+}
+
+func deletePorts(is *Service, object runtime.Object, nets []servers.Network) error {
+	for _, n := range nets {
+		if err := deletePort(is, object, n.Port); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func deleteAttachinterface(is *Service, object runtime.Object, instanceID, portID string) error {
+	instance, err := is.GetInstance(instanceID)
+	if err != nil {
+		return err
+	}
+
+	if instance == nil || instance.ID == "" {
+		return nil
+	}
+
+	port, err := ports.Get(is.networkClient, portID).Extract()
+	if err != nil {
+		if capoerrors.IsNotFound(err) {
+			return nil
+		}
+	}
+
+	err = attachinterfaces.Delete(is.computeClient, instanceID, portID).ExtractErr()
+	if err != nil {
+		if capoerrors.IsNotFound(err) {
+			return nil
+		}
+		record.Warnf(object, "FailedDeleteAttachinterface", "Failed to delete attachinterface: instance %s, port %s: %v", instance.ID, port.ID, err)
+		return err
+	}
+
+	record.Eventf(object, "SuccessfulDeleteAttachinterface", "Deleted attachinterface: instance %s, port %s", instance.ID, port.ID)
+	return nil
+}
+
+func deleteTrunk(is *Service, object runtime.Object, portID string) error {
+	port, err := ports.Get(is.networkClient, portID).Extract()
+	if err != nil {
+		if capoerrors.IsNotFound(err) {
+			return nil
+		}
+	}
```

Comment:
Same here: houldn't the error still get returned if it is not `IsNotFound`?
```suggestion
		return err
	}
```